compiler-rt/test/tsan/java_heap_init2.cpp

```diff
@@ -0,0 +1,32 @@
+// RUN: %clangxx_tsan -O1 %s -o %t && %run %t 2>&1 | FileCheck %s
```

Comment:
Please add these test cases to a separate pull request. Since your fix is not committed yet, the tests would need an ["XFAIL" annotation](https://llvm.org/docs/TestingGuide.html#constraining-test-execution). Then, after the test cases are landed, this fix PR would update the test by removing the XFAIL. This two-step procedure makes it clear that 1) TSan was broken 2) this PR fixes it.

Reply:
If it was unfixed in your environment, that was an exception (and would you mind providing more information about the test failure, such as the triggering environment?); 

This relevant bug should indeed be fixed by this PR, and I have passed this test (Ubuntu24.04@x86_64) before I pushed the commit.

FYI, this test case manages to trigger a situation that `munmap` clears the meta incompletely due to the incorrect usage of an inclusive meta end. However, this bug should be fixed in this PR, as follows:
```cpp
  // This round-up gives the exclusive meta end.
  uptr size_for_meta = RoundUp(addr + size, kMetaShadowCell) - addr;
  ctx->metamap.ResetRange(thr->proc(), addr, size_for_meta, true);
```

Reply:
I submitted the testcase and fix together in several PRs before, so I did the same this time. 
But your suggestion is really a better practice, and I have put forward a separate PR #145472. :)
After that PR is merged, I will rebase this PR.

Reply:
Thanks!